atomics/include/desul/atomics/Compare_Exchange_HIP.hpp

```diff
@@ -17,130 +18,61 @@ SPDX-License-Identifier: (BSD-3-Clause)
-                                      reinterpret_cast<unsigned int&>(compare),
-                                      reinterpret_cast<unsigned int&>(value));
-  return reinterpret_cast<T&>(return_val);
-}
-template <class T, class MemoryScope>
-__device__ std::enable_if_t<sizeof(T) == 8, T> device_atomic_compare_exchange(
-    T* const dest, T compare, T value, MemoryOrderRelaxed, MemoryScope) {
-  static_assert(sizeof(unsigned long long int) == 8,
-                "this function assumes an unsigned long long is 64-bit");
-  unsigned long long int return_val =
-      atomicCAS(reinterpret_cast<unsigned long long int*>(dest),
-                reinterpret_cast<unsigned long long int&>(compare),
-                reinterpret_cast<unsigned long long int&>(value));
-  return reinterpret_cast<T&>(return_val);
-}
+template <class T>
+struct atomic_exchange_available_hip {
+  constexpr static bool value =
+  ((sizeof(T) == 1 && alignof(T) == 1) || (sizeof(T) == 4 && alignof(T) == 4) ||
+   (sizeof(T) == 8 && alignof(T) == 8)) &&
+      std::is_trivially_copyable<T>::value;
+};
 
-template <class T, class MemoryScope>
-__device__ std::enable_if_t<sizeof(T) == 4 || sizeof(T) == 8, T>
-device_atomic_compare_exchange(
-    T* const dest, T compare, T value, MemoryOrderRelease, MemoryScope) {
-  T return_val = atomic_compare_exchange(
-      dest, compare, value, MemoryOrderRelaxed(), MemoryScope());
-  atomic_thread_fence(MemoryOrderRelease(), MemoryScope());
-  return return_val;
-}
-
-template <class T, class MemoryScope>
-__device__ std::enable_if_t<sizeof(T) == 4 || sizeof(T) == 8, T>
+template <class T, class MemoryOrder, class MemoryScope>
+__device__ std::enable_if_t<atomic_exchange_available_hip<T>::value, T>
 device_atomic_compare_exchange(
-    T* const dest, T compare, T value, MemoryOrderAcquire, MemoryScope) {
-  atomic_thread_fence(MemoryOrderAcquire(), MemoryScope());
-  T return_val = atomic_compare_exchange(
-      dest, compare, value, MemoryOrderRelaxed(), MemoryScope());
-  return return_val;
+    T* const dest, T compare, T value, MemoryOrder, MemoryScope) {
+  (void)__hip_atomic_compare_exchange_strong(
+      dest,
+      &compare,
+      value,
+      HIPMemoryOrder<MemoryOrder>::value,
+      HIPMemoryOrder<cmpexch_failure_memory_order<MemoryOrder>>::value,
+      HIPMemoryScope<MemoryScope>::value);
+  return compare;
 }
 
 template <class T, class MemoryScope>
-__device__ std::enable_if_t<sizeof(T) == 4 || sizeof(T) == 8, T>
-device_atomic_compare_exchange(
-    T* const dest, T compare, T value, MemoryOrderAcqRel, MemoryScope) {
-  atomic_thread_fence(MemoryOrderAcquire(), MemoryScope());
-  T return_val = atomic_compare_exchange(
-      dest, compare, value, MemoryOrderRelaxed(), MemoryScope());
-  atomic_thread_fence(MemoryOrderRelease(), MemoryScope());
+__device__ std::enable_if_t<atomic_exchange_available_hip<T>::value, T>
+device_atomic_exchange(T* const dest, T value, MemoryOrderRelaxed, MemoryScope) {
```

Comment:
Did you mean to keep that overload?

Reply:
yes

Reply:
Please explain why we need them and the more generic one is not enough.

Reply:
Because the signature is different. The one you call more "generic" take a `compare` operator. This function doesn't.

Reply:
Why are we not doing
```C++
template <class T, class MemoryOrder, class MemoryScope>
__device__ std::enable_if_t<atomic_exchange_available_hip<T>::value, T>
device_atomic_exchange(T* const dest, T value, MemoryOrderRelaxed, MemoryScope) {
  return __hip_atomic_exchange(dest,
                               value,
                               HIPMemoryOrder<MemoryOrder>::value,
                               HIPMemoryScope<MemoryScope>::value);
}

Reply:
We only support this API for a subset of the `MemoryOrder`. We could extent it but this PR just replace the implementation of the current API.

Reply:
> We only support this API for a subset of the `MemoryOrder`.

You mean the current overload set (before this PR) would not find a match for `MemoryOrderSeqCst`?

Please prefer the generic overload.  We'd probably want to handle constraints/mandates consistently.  We don;t need to handle it in this PR.  With the current proposed version, if trying to call seq_cst with a non-supported type that goes though the lock-based implementation would fail to compile with a possibly confusing error message.

Reply:
Sure, I'll do it. I just didn't want to add new functionality in a PR that was only supposed to change the implementation.

Reply:
I've added the generic implementation of `device_atomic_exchange`. I've also removed the `device_atomic_exchange` function that was taking an unused `compare` operator. This was only implemented for `MemoryOrderAcquire`. This function does not exist for the other backend and doesn't make sense too me. Why give a `compare` operator that is ignored when calling `device_atomic_exchange` when there is `device_atomic_compare_exchange` which uses the operator. Let me know if you want me to put this function back.